design-proposals/vhostuser-interface.md

```diff
@@ -0,0 +1,78 @@
+
+## Non Goals
+- The user space networking will require the host to a DPDK dataplane.
+- The kubevirtci creates kernel mode dataplane hosts for testing the VMs. It needs to be updated to support DPDK.
+
+## Definition of Users
+- Any user who has a (k8s) cluster with a dpdk dataplane can use the interface/feature.
+- The feature will be supported only if a valid CNI is being used.
+
+## User Stories
+- As a user/admin, I want to use user space networking feature if I have DPDK supported VM and the host has a DPDK dataplane.
+
+## Repos
+
+- kubevirt/kubevirt
+   - Introduction of vhostuser type interface.
+- kubevirt/kubevirtci
+   - Enable usage of dpdk on the setups created by kubevirtci.
+
+# Design
+- The design involves introducing vHostUser interface type and required parameters to support it. 
+
+- Creating the VMs will follow the same process with a few changes highlighted as below:
+    1. A client posts a new VM definition to the K8s API Server.  
+    2. The K8s API Server validates the input and creates a VM custom resource definition (CRD) object.  
+    3. The virt-controller observes the creation of the new VM object and creates a corresponding virt-launcher pod.
+    4. **The virt-controller will add two new volumes to the virt-launcher pod.**\
+        a. **EmptyDir volume named (shared-dir) is used to share the vhostuser socket file with the virt-launcher pod and dpdk dataplane.**\
+        b. **DownwardAPI volume named (podinfo) is used to share vhostuser socket file name with the virt-launcher via pod annotations.**
+    5. Scheduling of the launcher pod will be handled by kubernetes.
+    6. **The CNI will mount the shared-dir i.e., EmptyDir volume with the mount-path as "/var/lib/cni/usrcni/".**
+    7. **The CNI will update the virt-launcher pod annotations with vhostsocket-file name and details.**
+    8. The virt-controller observes the creation of the virt-launcher pod for the VM and updates the nodeName field in the VM object with the nodeName retrieved from the virt-launcher pod.
+    9. The virt-handler observes that a VM got assigned to host where it is running on.
+   10. The virt-handler will provide the VM Specification and signals the virt-launcher via IPC(unix domain socket) to create the VM in the virt-launcher pod.
+   11. The virt-launcher using a libvirtd instance creates the domain.
+   12. **The virt-launcher reads the DownwardAPI volume and retrieves the vhostsocket-file name specified in pod annotations and uses it to create a unix socket in the VM, while launching the VM using libvirtd.**
+   13. **The virt-launcher is modified to skip establishing the networking between VM and the virt-launcher pod using bridge(Refer in Kubevirt Networking section in https://kubevirt.io/2018/KubeVirt-Network-Deep-Dive.html).**
```

Comment:
How would you do that? Do we need a new network mode?

Reply:
The new interface will be using the vhost socket created by the shared-dir mount. This will allow the VM with a dpdk interface to connect to the DPDK data plane directly bypassing the linux.